File: scripts/check_deterministic_bundle.py
#!/usr/bin/env python3
"""Verify that the generated plugin bundle is reproducible from source."""
from __future__ import annotations

import hashlib
import subprocess
import sys
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
BUNDLER = ROOT / "scripts" / "bundle_obsidian_plugin.py"
GENERATED_BUNDLE = ROOT / "main.js"


def build_bundle(output: Path) -> bytes:
    subprocess.run(
        [
            sys.executable,
            str(BUNDLER),
            "--plugin-dir",
            str(ROOT / "src"),
            "--out",
            str(output),
        ],
        cwd=ROOT,
        check=True,
        stdout=subprocess.DEVNULL,
    )
    return output.read_bytes()


def main() -> None:
    if not GENERATED_BUNDLE.exists():
        raise SystemExit("main.js has not been generated; run npm run build first")

    with tempfile.TemporaryDirectory(prefix="understory-bundle-check-") as tmp:
        tmp_dir = Path(tmp)
        first = build_bundle(tmp_dir / "first.js")
        second = build_bundle(tmp_dir / "second.js")

    if first != second:
        raise SystemExit("Two clean bundle builds produced different bytes")
    if GENERATED_BUNDLE.read_bytes() != first:
        raise SystemExit("Generated main.js does not match a clean source build")

    digest = hashlib.sha256(first).hexdigest()
    print(f"Deterministic bundle OK: sha256={digest}")


if __name__ == "__main__":
    main()

File: scripts/check_deterministic_bundle.js
#!/usr/bin/env node
/* eslint-disable @typescript-eslint/no-require-imports, @typescript-eslint/no-unsafe-argument -- Node release verification script is intentionally CommonJS so it can validate the npm build entrypoint. */

const cryptoModule = require('crypto');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { spawnSync } = require('child_process');

const ROOT = path.resolve(__dirname, '..');
const BUNDLER = path.join(ROOT, 'scripts', 'bundle_obsidian_plugin.js');
const GENERATED_BUNDLE = path.join(ROOT, 'main.js');

function buildBundle(output) {
    const result = spawnSync(
        process.execPath,
        [
            BUNDLER,
            '--plugin-dir',
            path.join(ROOT, 'src'),
            '--out',
            output,
        ],
        { cwd: ROOT, stdio: ['ignore', 'ignore', 'inherit'] }
    );
    if (result.status !== 0) {
        throw new Error(`Bundle build failed with status ${result.status}`);
    }
    return fs.readFileSync(output);
}

function main() {
    if (!fs.existsSync(GENERATED_BUNDLE)) {
        throw new Error('main.js has not been generated; run npm run build first');
    }

    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'understory-bundle-check-'));
    try {
        const first = buildBundle(path.join(tmpDir, 'first.js'));
        const second = buildBundle(path.join(tmpDir, 'second.js'));

        if (!first.equals(second)) {
            throw new Error('Two clean bundle builds produced different bytes');
        }
        if (!fs.readFileSync(GENERATED_BUNDLE).equals(first)) {
            throw new Error('Generated main.js does not match a clean source build');
        }

        const digest = cryptoModule.createHash('sha256').update(first).digest('hex');
        process.stdout.write(`Deterministic bundle OK: sha256=${digest}\n`);
    } finally {
        fs.rmSync(tmpDir, { recursive: true, force: true });
    }
}

if (require.main === module) {
    main();
}

/* eslint-enable @typescript-eslint/no-require-imports, @typescript-eslint/no-unsafe-argument -- End Node release verification script bridge. */
